### ISSUE TITLE
Add suppport for Import of Scope Tags, Role Scope Tags

### ISSUE THREAD

**Issue by redkaffe** [nicolonsky/ModernWorkplaceConcierge#51]
Scope tags are already exported, and importing them could be a quick win if implemented.

**Comment by Kirikou974** [comment]
I have been looking at this and the solution doesn't look simple.
Example :
- Export is "easy", If I got it correctly, scope tags ids are used in DeviceConfiguration

> - I see that there in an array of roleScopeTagIds in a given DeviceConfiguration json :

```
  "roleScopeTagIds": [
    "4"
  ],
```

> - Scope tags are exported  as well and we can find the scope tag detail in the corresponding json file : 

```
{
  "displayName": "Scope tag display name",
  "description": "Scope tag description",
  "isBuiltIn": false,
  "id": "4",
  "@odata.type": "microsoft.graph.roleScopeTag"
}
```
- Import is another story : 
> - Scope tags would have to be imported first which is not the case right now. The code loops through the zip structure or folder (probably in alphabetical order ?) but this one is an easy fix IMHO
> - Scope tag id is automatically generated : https://docs.microsoft.com/fr-fr/graph/api/intune-rbac-rolescopetag-create?view=graph-rest-beta (id = Key of the entity. This is read-only and automatically generated) ==> this means the program would have to keep track of the old id and the corresponding new id
> - DeviceConfiguration would have to be imported after the scope tags and the program would need to match the linked scope tags ids with the old scope tag ids and then map those with the new scope tag ids

The way I see it, it implies 3 changes : 
- a rather minor one in the IntuneController.cs class to change the order of import (tags first)
- one in the GraphIntuneImport.cs class to add support for scope tag import and to find a neat way to map old and new ids
- and another one in GraphIntuneImport.cs class to add the link between scope tags and deviceconfigurations

Easier said than done though 😅

**Comment by enoola** [comment]
Hello Nicola,
I won't argue one the overall organisation (I'm not yet acquainted yet).
I spent quite a bit of time on finding how to create scope tags with graphapibeta, so here is my share if it can help.

In regards to importing RoleScopeTags as far as my attempts went you will need to steps :
1 : create the scope tag
2 : create the assignment (which makes sense looking at the fact rolescopetag can have multiple assignments

so to create it I use something like this (I am new to C#, so probably ugly but is is working):
Well in our project I don't need async method but looking at the quality of yours I am not afraid you know hwat you are doing :)

----Begin of RoleScopeTag creation creation ----

```
HttpClient client = new HttpClient();
client.DefaultRequestHeaders.Accept.Clear();
//szGraphApiRoleScopeTags = "/deviceManagement/roleScopeTags
            
client.DefaultRequestHeaders.Add("Authorization", "Bearer " + AuthenticationHelper.TokenForUser);
            string url = AuthenticationHelper.szGraphApiBaseURL + AuthenticationHelper.szGraphApiVersion + AuthenticationHelper.szGraphApiRoleScopeTags;

JObject oRoleScopeTag = new JObject
{
    ["@odata.type"] = "#microsoft.graph.roleScopeTag",
    ["displayName"] = _szRoleScopeTagName,
    ["description"] = _szRoleScopeTagDescription,
};

 string json = JsonConvert.SerializeObject(oRoleScopeTag);
 var data = new StringContent(json, Encoding.UTF8, "application/json");
 var response = client.PostAsync(url, data).GetAwaiter().GetResult();

```
----End of RoleScopeTag creation creation ----

----Begin of RoleScopeTagAssignment ----

```
HttpClient client = new HttpClient();
client.DefaultRequestHeaders.Accept.Clear();

client.DefaultRequestHeaders.Add("Authorization", "Bearer " + AuthenticationHelper.TokenForUser);

// "https://graph.microsoft.com/beta/deviceManagement/roleScopeTags/$($RoleScopeTagId)/assignments"
string url = AuthenticationHelper.szGraphApiBaseURL + AuthenticationHelper.szGraphApiVersion + AuthenticationHelper.szGraphApiRoleScopeTags +'/' + Id + "/assignments" ;

JObject oRoleScopeTag = GetJSonUglyObject(GroupId);

string json = JsonConvert.SerializeObject(oRoleScopeTag);

var data = new StringContent(json, Encoding.UTF8, "application/json");

var response = client.PostAsync(url, data).GetAwaiter().GetResult();
string szJsonResult = response.Content.ReadAsStringAsync().Result;
            
if (! response.IsSuccessStatusCode ) {
  WriteObject(szJsonResult);
   return;
}

JObject oJsonRoleScopeTagGroupAssignement = JObject.Parse(szJsonResult);
CommonStruct.RoleScopeTagGroupAssignmentTarget oResRoleScopeTagGroupAssignment = ConvertHelper(oJsonRoleScopeTagGroupAssignement);

```

----End of RoleScopeTagAssignment ----

hope it helps, even a bit :)

if you'd like I'll send you the whole files by email (might be more convenient), le me know.
John PIGERET

**Comment by nicolonsky** [comment]
We need a way to:

1. Verify all referenced scope tags from device configs in an export (check whether they exist)

    * Compare name and ID of scope tag

2. Create missing scope tags (needs to happen before importing device configs)

3. On device config import adjust scope tag assignments to the respective scope tag id

**Comment by nicolonsky** [comment]
Added support for scope tag import & functionality to reassign policies. Curious about your feedback @redkaffe @Kirikou974 @enoola

**Comment by redkaffe** [comment]
just testes it out!!!! nice and working like a charm...have to do some more tests after setting up (again) my second and third tenants......be back with more details, but import from ZIP works perfectly, great job! again! thanks :)
